examples/nemo/eg3/kernel_utils.py

```diff
@@ -0,0 +1,142 @@
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+
+''' Module containing various utilities to aid in the application of
+    OpenACC KERNELS directives to NEMO source. Mainly required to
+    workaround the vagaries of the PGI compiler's support for OpenACC.
+'''
+
+
+def valid_kernel(node):
+    '''
+    Whether the sub-tree that has `node` at its root is eligible to be
+    enclosed within an OpenACC KERNELS directive.
+
+    :param node: the node in the PSyIR to check.
+    :type node: :py:class:`psyclone.psyGen.Node`
+
+    :returns: True if the sub-tree can be enclosed in a KERNELS region.
+    :rtype: bool
+
+    '''
+    from psyclone.nemo import NemoIfBlock, NemoIfClause
+    from psyclone.psyGen import CodeBlock
+    from fparser.two.utils import walk_ast
+    from fparser.two import Fortran2003
+    # PGI (18.10) often produces code that fails at run time if a Kernels
+    # region includes If constructs.
+    excluded_nodes = (CodeBlock, NemoIfBlock, NemoIfClause)
+    if isinstance(node, excluded_nodes):
+        return False
+    code_blocks = node.walk(node.children, excluded_nodes)
+    if code_blocks:
+        return False
+    # Check that there are no derived-type references in the sub-tree (because
+    # PGI deep-copy doesn't like them).
+    if walk_ast([node.ast], [Fortran2003.Data_Ref]):
```

Comment:
As discussed, it's because we do not currently check the contents of a proposed `NemoKern` very thoroughly. This is addressed in #365. I've added a comment to the code.